src/main/java/websocket/WSConnection.java

```diff
@@ -0,0 +1,340 @@
+     */
+    void messageLoop() {
+        WSMessage msg;
+        while (true) {
+            synchronized (this) {
+                if (getState() != State.READY) {
+                    return;
+                }
+            }
+            synchronized (messageQueue) {
+                msg = messageQueue.poll();
+                if (msg == null) {
+                    try {
+                        messageQueue.wait();
+                    } catch (InterruptedException e) {
+                        // Do nothing; try again next time.
+                    }
+                    continue;
+                }
+            }
+            sendMessage(msg);
+        }
+    }
+
+    /**
+     * Sends the message through the websocket. Will exit without doing anything if state is CLOSE or EXIT.
+     *
+     * @param msg Message to be sent
+     * @return true if message was sent successfully, false otherwise.
+     * @throws IllegalStateException if no connection found.
+     */
+    boolean sendMessage(WSMessage msg) {
+        synchronized (this) {
+            if (state == State.CLOSE || state == State.EXIT) {
+                return false;
+            } else if (state != State.READY) {
+                throw new IllegalStateException("Cannot send message if not in ready state");
+            }
+        }
+        Future<Void> fut = session.getRemote().sendStringByFuture(msg.getMessage());
+        try {
+            fut.get(5, TimeUnit.SECONDS);
+            logger.debug(String.format("Sent message: %s", msg.getMessage()));
+        } catch (CancellationException | ExecutionException | InterruptedException | TimeoutException e) {
+            logger.warn(String.format("Error sending message \"%s\" - Exception: %s", msg.getMessage(), e.getCause().getMessage()));
+
+            // Insert into retry queue.
+            if (msg.getRetryCount() < config.getMaxRetryCount()) {
+                msg.incrementRetryCount();
+                synchronized (messageQueue) {
+                    this.messageQueue.offer(msg);
+                    this.messageQueue.notifyAll();
+                }
+            } else {
+                for (IMessageHandler handler : incomingMessageHandlers) {
+                    handler.handleMessageSendError(msg.getMessage());
+                }
+            }
+
+            return false;
+        }
+        return true;
+    }
+
+    void enqueueMessage(String msg) {
+        if (getState() == State.CLOSE || state == State.EXIT) {
+            return;
+        }
+
+        synchronized (messageQueue) {
+            this.messageQueue.offer(new WSMessage(msg));
+            this.messageQueue.notifyAll();
+            logger.debug(String.format("Enqueued message: %s", msg));
+        }
+
+    }
+
+    @SuppressWarnings("WeakerAccess")
+    @OnWebSocketMessage
+    public void onMessage(String msg) {
+        logger.debug(String.format("Received message: %s", msg));
+        for (IMessageHandler handler : incomingMessageHandlers) {
```

Comment:
Why have incomingMessageHandlers be a List? That makes you do a for loop through all of them, and because at this point msg isnt parsed and is still a raw string, that would mean each handler would have to parse it? Shouldn't we be parsing it before sending it to them and be sure we're only sending it to the correct handler? It doesn't make sense to have each handler decide if it should run.